nixos/modules/system/boot/loader/generic-extlinux-compatible/extlinux-conf-builder.sh

```diff
@@ -53,8 +55,9 @@ declare -A filesCopied
 copyToKernelsDir() {
     local src=$(readlink -f "$1")
     # Don't copy if on same filesystem, just point to the original.
-    if [[ $(stat -f -c %i $src) = $(stat -f -c %i $target/nixos) ]]; then
-        echo "$src"
+    if [[ -z "@copyKernels@" && $(stat -f -c %i $src) = $(stat -f -c %i $target/) ]]; then
```

Comment:
`-z` (True if the length of string is zero) is the wrong check here, I think, see above comment.

Double quotes are not required inside `[[` since word splitting does not apply there.

Also, the use of a trailing slash in `$src` and `$target/` is now inconsistent.

You merely take `copyKernels = true` as hint and still don't copy files if `$src` and `$target` are on the same device/filesystem.

This behaviour a) contradicts your option's description and b) is not the same as the GRUB2 one, which simply defaults `copyKernels` to `true` if devices differ:
https://github.com/NixOS/nixpkgs/blob/5823018b1b27b9675a51a84a7fc9cdd44327fa3e/nixos/modules/system/boot/loader/grub/install-grub.pl#L99-L104

Reply:
I believe the logic is correct.

Double-quotes aren't required for a variable substitution (with `$`), but I think they are required here since the substitution is being done by substituteAll, before bash sees the script. No quotes and copyKernels = false would produce `if [[ -z && $(stat ...` ,which I don't trust bash to interpret correctly.

The trailing slash is to emphasize that I want to ask about the filesystem inside $target and not the one containing it. It's not required though, I can remove it.

copyKernels = false is the hint. Cases:

copyKernels = false and same fs -> true && true -> take this branch -> do not copy
copyKernels = false and different fs -> true && false -> don't take this branch -> copy
copyKernels = true and same fs -> false && true -> '' -> copy
copyKernels = true and different fs -> false && false -> '' -> copy

The code doesn't change the value of any copyKernels variables (it's already been substituted out of the script) but it acts as if it was true. It's extra-confusing because this if branch is the don't-copy branch.